---
title: "Downloading_data_and_Preprocessing"
author: "Gunjan Dixit"
date: "12/02/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Downloading Data and reference 

## Download raw data from SRA using GNU parallel and SRA-toolkit

The raw data can be downloaded from Sequence Read Archive [SRA] using the 'prefetch' command from SRA toolkit. Give the SRA Accession Ids in sequence in the following command to download them in parallel using GNU parallel. Once you have the raw files, extract the fastq files from these archive files with the help of 'fastq-dump'. Before converting .sra to fastq, copy/move the .sra file from each folder to your current working directory. 

```{bash}
#Creat project directory to store raw data
PROJ_DIR=AS_analysis/
mkdir $PROJ_DIR
cd $PROJ_DIR

RAW_DATA=raw_data
mkdir -p $RAW_DATA
cd $RAW_DATA
# Download SRA files using accession numbers
seq 10261601 10261606 | parallel prefetch SRR{}
mv SRR*/*.sra . 
#Rename the files according to the mapping information from GEO entry of the dataset. The mapping file should contain two columns, the first one should be SRA Run numbers and the second column should contain sample names obtained from metadata.
awk -F',' 'system("mv " $1 " " $2)' mapping.txt
rm -rf SRR*
  
#Convert SRA to fastq
parallel -j 3 fastq-dump --gzip --skip-technical --read-filter pass --dumpbase --split-e --clip --origfmt {} ::: Mbnl1KO_Thymus_1 Mbnl1KO_Thymus_2 Mbnl1KO_Thymus_3 WT_Thymus_1 WT_Thymus_2 WT_Thymus_3 

#Renaming fastq files for the paired reads to be identified by R1 and R2
for i in *.fastq.gz; 
  do 
  mv -v "$i" "${i/_pass/R}"; 
done
```

## Download Reference genome for read alignment
The reference genome and annotations for Mouse (Genome assembly GRCm39) were downloaded from www.ensembl.org.

```{bash}
cd ../
wget -nv -O annotation.gtf.gz http://ftp.ensembl.org/pub/release-103/gtf/mus_musculus/Mus_musculus.GRCm39.103.gtf.gz \
    && gunzip -f annotation.gtf.gz
wget -nv -O genome.fa.gz http://ftp.ensembl.org/pub/release-103/fasta/mus_musculus/dna/Mus_musculus.GRCm39.dna.primary_assembly.fa.gz \
&& gunzip -f genome.fa.gz
GTF=$(readlink -f annotation.gtf)
GENOME=$(readlink -f genome.fa)
```

### Preprocessing
Preprocessing includes two major steps-
1. Quality Check
2. Read Alignment

## Quality Control using FASTQC
FASTQC is the most widely used tool to assess the quality of reads generated by high throughput sequencing technologies. We create the output folder and then run quality check on all the fastq files together using GNU parallel. 

```{bash}
#Quality control using FASTQC
mkdir fastqc_out
parallel "fastqc {} -o fastqc_out" ::: $RAW_DATA/*.fastq.gz
```

## Read alignment

The next step in preprocessing includes mapping the reads to reference genome. We first build the index for reference genome and then use read pairs to align them. STAR aligner was used for this purpose. 

```{bash}
#Build STAR index
GDIR=STAR_indices
mkdir $GDIR
STAR --runMode genomeGenerate --genomeFastaFiles $GENOME --sjdbGTFfile $GTF --runThreadN 8 --genomeDir $GDIR

ODIR=results/mapping
mkdir -p $ODIR

#Align reads to genome
for fq1 in $RAW_DATA/*R1.fastq.gz; 
do
fq2=$(echo $fq1 | sed 's/1.fastq.gz/2.fastq.gz/g');
OUTPUT=$(basename ${fq1}| sed 's/R1.fastq.gz//g');

STAR --genomeDir $GDIR \
 --runThreadN 12 \
 --readFilesCommand zcat \
 --readFilesIn  ${fq1} ${fq2}\
 --outFileNamePrefix $ODIR\/${OUTPUT} \
 --outSAMtype BAM SortedByCoordinate \
 --outSAMunmapped Within \
 --outSAMattributes Standard

done

#Rename bam files and move to another directory
for bm in $ODIR/*.bam; 
  do 
    mv -v "$bm" "${bm/_Aligned.sortedByCoord.out.bam/.bam}"; 
done;
mkdir bams
mv $ODIR/*.bam bams/ 
```

Once the reads are mapped, alignment stats/metrics can be checked to identify the percentage of uniquely mapped reads. Higher the number, better is the alignment rate. After this we need to prepare/process annotation file for reference genome to incorporate exon data. Refer to "Prepare_mm10_exon_annotation.R' script for further analysis.  
